core/services/run_queue.go

```diff
@@ -34,81 +35,180 @@ type RunQueue interface {
 }
 
 type runQueue struct {
-	workersMutex  sync.RWMutex
-	workers       map[string]int
-	workersWg     sync.WaitGroup
-	stopRequested bool
+	runExecutor  RunExecutor
+	workers      map[models.ID]*singleJobSpecWorker // Access to this field is serialized through the select in orchestrateWorkers().
+	numWorkers   int32                              // Note: every access to this field must be atomic, as it is shared between goroutines
```

Comment:
Isn't this exactly what a WaitGroup is supposed to do?

Reply:
A WaitGroup could've been used in place of the top-level `runQueue.chDone` channel, but I couldn't see a compelling reason to mix concurrency primitives here.